sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1521,13 +1559,63 @@ static const bool FilterEventWaitList = [] {
 }();
 
 pi_result _pi_ze_event_list_t::createAndRetainPiZeEventList(
-    pi_uint32 EventListLength, const pi_event *EventList, pi_queue CurQueue) {
+    pi_uint32 EventListLength, const pi_event *EventList, pi_queue CurQueue,
+    bool UseCopyEngine) {
+  auto &LastCommandEvent = CurQueue->LastCommandEvent;
+  bool UseLastEvent = true;
+
+  if (CurQueue->isEventlessMode()) {
+    // In eventless mode, to support in-order semantics, it adds a barrier or
+    // barrier with an event to the previous command-list if the list is open.
+    auto &CopyCommandListIt = CurQueue->CopyCommandBatch.OpenCommandList;
+    auto &ComputeCommandListIt = CurQueue->ComputeCommandBatch.OpenCommandList;
+    auto &LastCommandList =
+        (CopyCommandListIt != CurQueue->CommandListMap.end() &&
+         !CopyCommandListIt->second.EventList.empty() &&
+         LastCommandEvent == CopyCommandListIt->second.EventList.back())
+            ? CopyCommandListIt
+            : ComputeCommandListIt;
+    if (LastCommandList != CurQueue->CommandListMap.end()) {
+      auto &CurrentCommandList =
+          UseCopyEngine ? CopyCommandListIt : ComputeCommandListIt;
+      if (LastCommandList != CurrentCommandList) {
```

Comment:
I think we need to add "in-order" bookkeeping for any switch of command-list, i.e. without seeing if it is copy or not. And then, I believe, you don't need the new `UseCopyEngine` argument.